### PR TITLE
WIP: Eliminate catastrophic backtracking that hangs Atom

### PR DESCRIPTION
**Pull request by nathansobo**
Closes https://github.com/atom/atom/issues/15030
Refs https://github.com/atom/atom/issues/14856 (Not sure if this also solves this one, but it might)

#99 introduced a regression that causes the parser to hang on certain kinds of diabolical input.

The first two commits of this PR introduce a minimal test case that isolates the problem. The third commit comments out the offending component of the regular expression that causes the problem.

@Alhadis @50Wliu Can we use a different approach to support the same functionality without this regular expression that has a potential for catastrophic performance?

I'm not very well versed in these grammars, but I'm inclined to revert #99 and go back to the older, less capable grammar if we can't guarantee good performance. Let me know if I should wait for you to attempt a fix or go ahead and revert.

Thank you.

/cc @iolsen @ungb

### CODE REVIEW

File: grammars/css.cson [hunk 1]
```diff
@@ -1913,7 +1913,8 @@
             'match': '''(?x)
               (-?(?!\\d)(?:[\\w-]|[^\\x00-\\x7F]|\\\\(?:[0-9a-fA-F]{1,6}|.))+)
               \\s*
-              (?=[~|^\\]$*=]|/\\*)
```
Comment (Alhadis) [inline grammars/css.cson]: What the HELL, that quantifier shouldn't be there. :| Oops.

@nathansobo Try removing `$*`, that's a zero-length match that'll match anything. Really bad news.

Comment (Alhadis) [inline grammars/css.cson]: In fact, I bet dollars to donuts that's what's causing the regression.

Comment (nathansobo) [inline grammars/css.cson]: Thanks for your prompt reply. Watching my daughter so I'll give it a try later. Thanks again.

**Comment by 50Wliu** [comment]
Thanks for taking this on @nathansobo and sorry I didn't get to it sooner 😬.

File: grammars/css.cson (hunk 1 above)
Comment (50Wliu) [inline grammars/css.cson]: Removing the asterisk does not fix the issue.

**Comment by 50Wliu** [comment]
> I'm inclined to revert #99 and go back to the older, less capable grammar if we can't guarantee good performance.

Unfortunately, that's more involved as it sounds as language-sass and language-less now rely on the new patterns introduced in #99.

**Comment by Alhadis** [comment]
@50Wliu Oh right, whoops. Still waking up, haha, sorry.

I'd recommend grouping the embedded grammar inside something that tries the "clamping" technique, but @50Wliu seems to think that's too "weird", so I don't know what to do. =)

File: grammars/css.cson (hunk 1 above)
Comment (50Wliu) [inline grammars/css.cson]: About to eat, but changing `(?:[\\w-]...)` to an atomic group `(?>[\\w-]...)` works.  Didn't run specs though.

Comment (Alhadis) [inline grammars/css.cson]: @nathansobo I'll wait for you (or somebody else) to confirm this before attempting a different solution. Trying to tie up loose ends elsewhere ATM.

Comment (nathansobo) [inline grammars/css.cson]: Removing the `*` after the `$` didn't seem to help. 😞

**Comment by nathansobo** [comment]
So what's this "clamping" technique you speak of? Can you link me to a description?

**Comment by Alhadis** [comment]
"Clamping" is a term I penned to describe a technique to prevent runaway issues with `begin`/`end` pattern-pairs. Embedding a pattern-set within a fixed-range (matched by `pattern`, as opposed to `begin/end`) carries no risk of an unclosed tag affecting the rest of the document.

... eh, hang on, I'll slap together a crude example so you can see what I mean.

**Comment by Alhadis** [comment]
Basically, it's an irregular and slightly messy (albeit highly effective) hack that we should only consider if an atomic group doesn't solve our problem.

**Comment by 50Wliu** [comment]
Ok, all specs pass with the atomic group, including the new regression test.

**Comment by nathansobo** [comment]
@50Wliu Do you want to push to this PR?

**Comment by 50Wliu** [comment]
Pushed!  I can also confirm that this fixes atom/atom#14856.

Regarding the spec, is there any reason to include the HTML grammar (L3257)?

**Comment by nathansobo** [comment]
> Regarding the spec, is there any reason to include the HTML grammar

No. I started with the assumption that it was an interaction between HTML and CSS and then realized it wasn't. So that line can go.

**Comment by Alhadis** [comment]
Good news regarding the simpler solution. ;-)

Just for posterity and reference, I'll post an example of what I was talking about. Compare [this](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-text&grammar_url=&grammar_text=%22a%22%3A+%22b%22%0D%0Aname%3A+%22Clamping+Example%22%0D%0AscopeName%3A+%22source.clamping%22%0D%0Apatterns%3A+%5B%0D%0A%09name%3A+%22markup.inserted.diff%22%0D%0A%09begin%3A+%22%3Ctag%3E%22%0D%0A%09end%3A+%22%3C%2Ftag%3E%22%0D%0A%09patterns%3A+%5B%7B%0D%0A%09%09name%3A+%22markup.deleted.diff%22%0D%0A%09%09begin%3A+%22%23css%7B%22%0D%0A%09%09end%3A+%22%7D%22%0D%0A%09%7D%5D%0D%0A%5D&code_source=from-text&code_url=&code=Text+%3Ctag%3E+%23css%7B+property%3A+value%3B+%7D+%3C%2Ftag%3E+Text%0D%0AText+%3Ctag%3E+%23css%7B+property%3A+value%3B+%3C%2Ftag%3E+Text%0D%0A) with [this](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-text&grammar_url=&grammar_text=%22a%22%3A+%22b%22%0D%0Aname%3A+%22Clamping+Example%22%0D%0AscopeName%3A+%22source.clamping%22%0D%0Apatterns%3A+%5B%0D%0A%09name%3A+%22markup.inserted.diff%22%0D%0A%09match%3A+%22%3Ctag%3E%28.*%29%3C%2Ftag%3E%22%0D%0A%09captures%3A%0D%0A%09%091%3A+patterns%3A+%5B%7B%0D%0A%09%09%09name%3A+%22markup.deleted.diff%22%0D%0A%09%09%09begin%3A+%22%23css%7B%22%0D%0A%09%09%09end%3A+%22%7D%22%0D%0A%09%09%7D%5D%0D%0A%5D&code_source=from-text&code_url=&code=Text+%3Ctag%3E+%23css%7B+property%3A+value%3B+%7D+%3C%2Ftag%3E+Text%0D%0AText+%3Ctag%3E+%23css%7B+property%3A+value%3B+%3C%2Ftag%3E+Text%0D%0A). The second example is using "clamping" to embed the pattern with runaway potential inside a precalculated range.

**Comment by nathansobo** [comment]
Thanks to both for engaging so rapidly. This went much smoother than I expected. ✨s to both of you.

**Comment by 50Wliu** [comment]
@nathansobo Done with my changes, feel free to merge when you're ready!

**Pull request merged**